### PR TITLE
Appsilon.develop

### PR DESCRIPTION
**Pull request by krystian8207**

### CODE REVIEW

File: R/assertions.R [hunk 1]
```diff
@@ -81,6 +93,12 @@ assert <- function(data, predicate, ..., success_fun=success_continue,
 
   if(length(keeper.vars)==0)
     stop("assert requires columns to be selected. Check number of arguments", call.=FALSE)
+
+  if(isTRUE(attr(data, "assertr_data_defective")))
+    return(defect_fun(
+      attr(data, "assertr_errors"), data, "assert", name.of.predicate,
+      as.character(dplyr::enexprs(...)), NA, description))
```
Comment (dokato) [inline R/assertions.R:100]: add `@import dplyr` to docs

File: R/utils.R [hunk 2]
```diff
@@ -79,3 +79,35 @@ has_all_names <- function(...){
...
+#'
+#' @export
+has_class <- function(..., class){
+  check_this <- list(...)
+  parent <- parent.frame()
+  given_classes <- lapply(check_this, function(name) class(parent$.top_env[[name]]))
+  all(given_classes %in% class)
+}
+
+generate_id <- function() {
```
Comment (dokato) [inline R/utils.R:120]: Short docs for that, please?

File: tests/testthat/test-assertions.R [hunk 3]
```diff
@@ -161,6 +228,16 @@ test_that("verify works with long predicates (fix #80)", {
   )
 })
 
+test_that("skip_chain_opts doesn't affect functionality outside chain for verify", {
+  expect_equal(verify(mtcars, mtcars$drat > 2, skip_chain_opts=TRUE), mtcars)
+  expect_equal(verify(mtcars, nrow(mtcars) > 30, skip_chain_opts=TRUE), mtcars)
+  expect_true(verify(mtcars, mtcars$drat > 2, success_fun=success_logical, skip_chain_opts=TRUE))
+  expect_true(verify(mtcars, nrow(mtcars) > 30, success_fun=success_logical, skip_chain_opts=TRUE))
+  expect_equal(verify(mtcars, drat > 3, error_fun = error_logical, skip_chain_opts=TRUE), FALSE)
```
Comment (dokato) [inline tests/testthat/test-assertions.R]: Minor thing: keep consistent space around `=`. Compare with the line above.

File: R/predicates.R [hunk 4]
```diff
@@ -181,6 +181,52 @@ in_set <- function(..., allow.na=TRUE, inverse=FALSE){
...
+#'
+#' not_in_set(3,4)(5)
+#'
+#' @export
+not_in_set <- function(..., allow.na=TRUE, inverse=FALSE){
+  the_call <- deparse(sys.call())
+  set <- c(...)
+  if(!length(set)) stop("can not test for membership in empty set")
+  fun <- function(x){
+    if(is.null(x))       stop("nothing to check set membership to")
```
Comment (dokato) [inline R/predicates.R]: indent or decrease nr of spaces

File: R/utils.R [hunk 5]
```diff
@@ -79,3 +79,35 @@ has_all_names <- function(...){
...
+  paste0(sample(c(LETTERS, letters, 0:9), 20, TRUE), collapse = "")
```
Comment (dokato) [inline R/utils.R]: I know it's extremely unlikely for this to be the same but for real uniqness, we could use eg date in secs added or sth?

File: R/assertions.R [hunk 6]
```diff
@@ -330,9 +394,15 @@ insist <- function(data, predicate_generator, ...,
...
-    return(success_fun(data))
+  # if all checks pass in current assertion
+  if(all(log.mat))
+    return(success_fun(data, "insist", name.of.predicate.generator, colnames(sub.frame), NA, description))
+
+  # if errors occured and verification was obligatory
+  if(obligatory)
+    attr(data, "assertr_data_defective") <- TRUE
+
+  assertion.id <- generate_id()
```
Comment (dokato) [inline R/assertions.R:405]: I'm pretty confused with naming convention. I don't have strong preferences to any, but until this point I assumed that dotted (`a.b`) is for functions and snaked (`a_b`) for variables, but it doesn't seem to be a case here ?

**Review by dokato — CHANGES_REQUESTED** [review]
Just a handful of minor remarks! Good job, impressive changes :)

File: R/assertions.R (hunk 1 above)
Comment (krystian8207) [reply R/assertions.R:100]: It's already included in the DESCRIPTION file (inside Imports), so it's not necessary in here.

File: R/utils.R (hunk 5 above)
Comment (krystian8207) [reply R/utils.R]: Added `Sys.time()` with milliseconds. Now we're 100% sure the IDs are different.

File: R/assertions.R (hunk 6 above)
Comment (krystian8207) [reply R/assertions.R:405]: I tried to find out naming convention in this package. The things I realized are:
* not exported functions and S3 methods are dotted
* exported functions are snaked
* variables passed (or the one that supposes to be passed) to any of above type of methods inherits the style from the method.

In example in here assertion.id is passed to  `make.assertr.assert.error` (that is not exported function) and all parameters passed to it are dotted.

**Review by dokato — APPROVED** [review]
LGTM

**Pull request merged**